lib/kcerts/ssh.go

```diff
@@ -0,0 +1,125 @@
+package kcerts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string) error {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return err
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if _, err := os.Stat(sshDir); os.IsNotExist(err) {
+		fmt.Printf("ssh directory %s does not exist, attempting to create it", sshDir)
+		err = os.Mkdir(sshDir, 0700)
+		if err != nil {
+			return fmt.Errorf("could not create dir %s err: %w", sshDir, err)
+		}
+	}
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	if _, err := os.Stat(knownHosts); os.IsNotExist(err) {
+		fmt.Printf("ssh authorized hosts file %s does not exist, creating it now \n", knownHosts)
+		err = os.Mkdir(knownHosts, 0644)
+		if err != nil {
+			return fmt.Errorf("could not create file %s: %w", knownHosts, err)
+		}
+	}
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	existingKnownHostsContent, err := ioutil.ReadFile(knownHosts)
```

Comment:
Now: if the read fails here, it can be for many different reasons. However, the file is optional: it's ok if it doesn't exist, ssh may have never created it (fresh installation). I would change the condition to only fail if the error is != than does not exist:

```
if err != nil && !os.IsNotExist(err) {
  return fmt.Errorf("...")
}
```
and...

Reply:
Implemented, but earlier in the code we do guarantee that it exists and has been made